Add method to get values from config tool based dependency

dcbaker opened the pull request:
What I want is to bet able to get the includepath from LLVM. What ended up happening was a lot of refactoring to get there.

This series adds a new ConfigToolDependency class, which acts as a share base class for for config tool dependencies like llvm, wxwidgets, and gnustep. It also allows this class to be used in factory type constructors (like cups and pcap). Among the features this adds is unified argument detection, support for multiple version comparisons, and a unified config-tool method name.

This is based on my previous llvm patches (that are waiting for merge already), because this code allows the LLVMDependency to be simplified further.

ignatenkobrain commented on mesonbuild/dependencies/ui.py:
```diff
@@ -23,12 +23,16 @@
 from .. import mlog
 from .. import mesonlib
-from ..mesonlib import MesonException, Popen_safe, version_compare, extract_as_list
+from ..mesonlib import (
```
__[flake8]__
* [F401] '..mesonlib.version_compare' imported but unused

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/dev.py:
```diff
@@ -129,101 +133,96 @@ class LLVMDependency(ExternalDependency):
+            # to figure out for ourselves what to link with. We'll do that by
+            # checking in the directory provided by --libdir for a library
+            # called libLLVM-<ver>.(so|dylib|dll)
+            libs = self.get_config_value(['--libdir'], 'link_args')[0]
```
__[flake8]__
* [F841] local variable 'libs' is assigned to but never used

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/dev.py:
```diff
@@ -129,101 +133,96 @@ class LLVMDependency(ExternalDependency):
+            expected_name = 'libLLVM-{}'.format(self.version)
+            re_name = re.compile(r'{}.(so|dll|dylib)'.format(expected_name))
+
+            for file_ in os.listdir(libdir):
```
__[flake8]__
* [F821] undefined name 'libdir'

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/dev.py:
```diff
@@ -129,101 +133,96 @@ class LLVMDependency(ExternalDependency):
+
+            for file_ in os.listdir(libdir):
+                if re_name.match(file_):
+                    self.link_args = ['-L{}'.format(libdir),
```
__[flake8]__
* [F821] undefined name 'libdir'

<!-- Comment by [SideCI](https://sideci.com) -->

ignatenkobrain commented on mesonbuild/dependencies/base.py:
```diff
@@ -164,6 +173,126 @@ def get_compiler(self):
+        # with the tools and tool_name class attributes set, this class is then
+        # instantiated and returned. The reduce function (method) is also
+        # attached, since python's pickle module won't be able to do anything
+        # with this dynamically generated class otherwise. 
```
__[flake8]__
* [W291] trailing whitespace

<!-- Comment by [SideCI](https://sideci.com) -->

dcbaker commented:
I've rebased this on master, added libwmf, added a deprecation warning for the old per-tool method arguments, and fixed the libwmf test.

I expect LLVM to fail on this without #2595

dcbaker commented:
the build failures seem to be CI failures not failures with this series.

jpakkane commented on test cases/common/163 config tool variable/meson.build:
```diff
@@ -0,0 +1,31 @@
+  error('MESON_SKIP_TEST LLVM not installed.')
+endif
+
+inc = dep_llvm.get_configtool_variable('--includedir')
```
Should this perhaps be `includedir` as that is its actual name?

jpakkane commented:
Overall this seems like a nice improvement. Now ⚔️ conflicts, so needs rebasing and something to be done to the review comment about variable names.

dcbaker commented:
Anything missing to merge this? I need the get_configtool_variable method for a couple of components in mesa.

jpakkane commented:
Asking again, why have the double dashes in the variable name to get? It would indicate a command line option rather than a variable name.

dcbaker commented:
Ultimately each of these tools is a one-off solution, so the only way to get variables is to pass a command line argument, since different tools provide different options.
Are you asking for meson to add the `--` for the caller? I can do that.

jpakkane commented:
It would be nice for the interface to be consistent. Currently for pkgconfig variables you'd do

    dep.get_pkgconfig_variable('varname')

but here it would be:

    dep_llvm.get_configtool_variable('--includedir')

We might even in the future add a method called, say, `get_underlying_variable` or something that forwards the call to whatever the correct method name may be.

dcbaker commented:
I've updated with the requested change to `get_configtool_variable`, so now it doesn't take -- arguments, it adds that itself.
I've also updated the test case to no have a collision and added `get_configtool_variable` to the reference manual.

dcbaker commented:
@jpakkane, is this looking good, or would you like to see some more changes? I'd really like to see this land for 0.44, and I'm setting aside time to make any changes you need to see to get this landed.
